File: Programmers/Javascript/level1/실패율.js
var N=15;
var stages=[12,12,12,12,12,6,6]

function solution(N, stages) {
    var all=stages.length
    var temp=[]
    for(var i=1;i<=N;i++)
        {
            if(all!=0){
                var ct=count(i,stages);
                temp.push([i,ct/all])
                all-=ct
            } else{
              temp.push([i,0])
            }

        }
    console.log(temp)
    var answer=temp.sort(function(x,y){if(x[1]===y[1]){return 1;}
    return y[1]-x[1]})
    console.log(answer)
    var list=[];
    for(var i of answer){
      list.push(i[0])
    }
    return list
}

console.log(solution(N,stages))

function count(num,list){
    var count=0
    for(var i of list)
        {
            if(i===num){
                count+=1
            }
        }
    return count
}

File: Programmers/Java11/src/com/programmers/실패율.java
import java.util.*;

public class 실패율 {
    static class Solution {
        public int[] solution(int N, int[] stages) {

            List<Stage> lastStages = new ArrayList<>();
            int total = stages.length;

            for (int i = 1; i < N+1; i++) {
                double cnt = 0;
                for (int j = 0; j < stages.length; j++) {
                    if (stages[j] == i) {
                        cnt+=1;
                    }
                }
                Double failRate = cnt / total;
                total -= cnt;

                Stage s = new Stage(i, failRate);
                lastStages.add(s);
            }

            Collections.sort(lastStages);


            int[] answer = new int[N];
            for (int i = 0; i < N; i++) {
                answer[i] = lastStages.get(i).id;
            }

            return answer;
        }

        class Stage implements Comparable<Stage> {
            public int id;
            public double failure;

            public Stage(int id_, double failure_) {
                id = id_;
                failure = failure_;
            }

            @Override
            public int compareTo(Stage o) {
                if (failure < o.failure ) {
                    return 1;
                }
                if (failure > o.failure ) {
                    return -1;
                }
                return 0;
            }
        }

    }

    public static void main(String[] args) {

        int[] stages = {2, 1, 2, 6, 2, 4, 3, 3};
        int N = 5;

        실패율.Solution solution = new 실패율.Solution();

        int[] solution1 = solution.solution(N, stages);
        for (int i : solution1) {
            System.out.println(i);
        }
    }
}
